spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java

```diff
@@ -0,0 +1,132 @@
+import org.springframework.amqp.core.Address;
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.core.ChannelAwareMessageListener;
+
+import com.rabbitmq.client.Channel;
+
+/**
+ * Listener container for Direct ReplyTo only listens to the pseudo queue
+ * {@link Address#AMQ_RABBITMQ_REPLY_TO} with a single consumer.
+ *
+ * @author Gary Russell
+ * @since 2.0
+ *
+ */
+public class DirectReplyToMessageListenerContainer extends DirectMessageListenerContainer {
+
+	private final ConcurrentMap<Channel, Boolean> inUseConsumerChannels = new ConcurrentHashMap<>();
+
+	private int consumerCount;
+
+	public DirectReplyToMessageListenerContainer(ConnectionFactory connectionFactory) {
+		super(connectionFactory);
+		setQueueNames(Address.AMQ_RABBITMQ_REPLY_TO);
+		setAcknowledgeMode(AcknowledgeMode.NONE);
+		super.setConsumersPerQueue(0);
+	}
+
+	@Override
+	public final void setConsumersPerQueue(int consumersPerQueue) {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public final void setMonitorInterval(long monitorInterval) {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public final void setQueues(Queue... queues) {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public final void addQueueNames(String... queueNames) {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public final void addQueues(Queue... queues) {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public final boolean removeQueueNames(String... queueNames) {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public final boolean removeQueues(Queue... queues) {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public void setChannelAwareMessageListener(ChannelAwareMessageListener messageListener) {
+		super.setChannelAwareMessageListener((message, channel) -> {
+			this.inUseConsumerChannels.remove(channel);
+			messageListener.onMessage(message, channel);
+		});
+	}
+
+	@Override
+	protected void doStart() throws Exception {
+		if (!isStarted()) {
+			this.consumerCount = 0;
+			super.setConsumersPerQueue(0);
+			super.doStart();
+		}
+	}
+
+	@Override
+	protected void consumerRemoved(SimpleConsumer consumer) {
+		this.inUseConsumerChannels.remove(consumer);
+	}
+
+	/**
+	 * Get the channel associated with a direct reply-to consumer.
+	 * If the consumer has exited, the container will be stopped.
+	 * @return the channel or null if there is no consumer.
+	 */
+	public Channel getChannel() {
+		synchronized (this.consumersMonitor) {
+			Channel channel = null;
+			while (channel == null) {
+				for (SimpleConsumer consumer : getConsumers()) {
+					if (this.inUseConsumerChannels.putIfAbsent(consumer.getChannel(), Boolean.TRUE) == null) {
+						channel = consumer.getChannel();
+						break;
+					}
+				}
+				if (channel == null) {
+					this.consumerCount++;
```

Comment:
The count is never decremented; we only remove them from the in use map - the monitor thread will recover any consumers that die.